--- a/arch/arm/boot/dts/sun7i-a20-bananapi.dts	2016-02-16 16:07:43.055787291 +0100
+++ a/arch/arm/boot/dts/sun7i-a20-bananapi.dts	2016-02-16 16:09:29.161111966 +0100
@@ -301,14 +301,14 @@
 	status = "okay";
 };
 
-&spi0 {
-	pinctrl-names = "default";
-	pinctrl-0 = <&spi0_pins_a>,
-		    <&spi0_cs0_pins_a>,
-		    <&spi0_cs1_pins_a>;
-	status = "okay";
-};
-
+/* &spi0 {
+ *	pinctrl-names = "default";
+ *	pinctrl-0 = <&spi0_pins_a>,
+ *		    <&spi0_cs0_pins_a>,
+ *		    <&spi0_cs1_pins_a>;
+ *	status = "okay";
+ *};
+ */
 &uart0 {
 	pinctrl-names = "default";
 	pinctrl-0 = <&uart0_pins_a>;
