diff --git a/trunk/can-utils/Makefile b/trunk/can-utils/Makefile
index 0fd201a..c027bfe 100644
--- a/trunk/can-utils/Makefile
+++ b/trunk/can-utils/Makefile
@@ -62,7 +62,6 @@ PROGRAMS = can-calc-bit-timing candump cansniffer cansend canplayer cangen canbu
 	   $(PROGRAMS_ISOTP)\
 	   $(PROGRAMS_CANGW)\
 	   $(PROGRAMS_SLCAN)\
-	   $(PROGRAMS_J1939)\
 	   slcanpty canfdtest
 
 all: $(PROGRAMS)
diff --git a/trunk/can-utils/cangen.c b/trunk/can-utils/cangen.c
index 769fac2..21163d4 100644
--- a/trunk/can-utils/cangen.c
+++ b/trunk/can-utils/cangen.c
@@ -64,7 +64,7 @@
 #include <linux/can/raw.h>
 #include "lib.h"
 
-#define DEFAULT_GAP 200 /* ms */
+#define DEFAULT_GAP (1000000)
 
 #define MODE_RANDOM	0
 #define MODE_INCREMENT	1
@@ -79,8 +79,8 @@ void print_usage(char *prg)
 {
 	fprintf(stderr, "\n%s: generate CAN frames\n\n", prg);
 	fprintf(stderr, "Usage: %s [options] <CAN interface>\n", prg);
-	fprintf(stderr, "Options: -g <ms>       (gap in milli seconds "
-		"- default: %d ms)\n", DEFAULT_GAP);
+	fprintf(stderr, "Options: -g <us>       (gap in microseconds "
+		"- default: %d us)\n", DEFAULT_GAP);
 	fprintf(stderr, "         -e            (generate extended frame mode "
 		"(EFF) CAN frames)\n");
 	fprintf(stderr, "         -f            (generate CAN FD CAN frames)\n");
@@ -133,7 +133,7 @@ void sigterm(int signo)
 
 int main(int argc, char **argv)
 {
-	double gap = DEFAULT_GAP;
+	unsigned int  gap = DEFAULT_GAP;
 	unsigned long polltimeout = 0;
 	unsigned char ignore_enobufs = 0;
 	unsigned char extended = 0;
@@ -163,7 +163,6 @@ int main(int argc, char **argv)
 	int i;
 	struct ifreq ifr;
 
-	struct timespec ts;
 	struct timeval now;
 
 	/* set seed value for pseudo random numbers */
@@ -182,7 +181,7 @@ int main(int argc, char **argv)
 			break;
 
 		case 'g':
-			gap = strtod(optarg, NULL);
+			gap = strtoul(optarg, NULL, 0);
 			break;
 
 		case 'e':
@@ -277,9 +276,6 @@ int main(int argc, char **argv)
 		return 1;
 	}
 
-	ts.tv_sec = gap / 1000;
-	ts.tv_nsec = (long)(((long long)(gap * 1000000)) % 1000000000ll);
-
 	/* recognize obviously missing commandline option */
 	if (id_mode == MODE_FIX && frame.can_id > 0x7FF && !extended) {
 		printf("The given CAN-ID is greater than 0x7FF and "
@@ -458,8 +454,7 @@ resend:
 		}
 
 		if (gap) /* gap == 0 => performance test :-] */
-			if (nanosleep(&ts, NULL))
-				return 1;
+			usleep(gap);
 
 		if (id_mode == MODE_INCREMENT)
 			frame.can_id++;
diff --git a/trunk/can-utils/jacd.c b/trunk/can-utils/jacd.c
index bac4060..b333f4e 100644
--- a/trunk/can-utils/jacd.c
+++ b/trunk/can-utils/jacd.c
@@ -19,7 +19,6 @@
 
 #include <unistd.h>
 #include <getopt.h>
-#include <error.h>
 #include <sys/time.h>
 #include <sys/socket.h>
 #include <net/if.h>
diff --git a/trunk/can-utils/jspy.c b/trunk/can-utils/jspy.c
index 55ddeb7..6edeb81 100644
--- a/trunk/can-utils/jspy.c
+++ b/trunk/can-utils/jspy.c
@@ -18,7 +18,6 @@
 
 #include <unistd.h>
 #include <getopt.h>
-#include <error.h>
 #include <sys/socket.h>
 #include <sys/ioctl.h>
 #include <sys/time.h>
diff --git a/trunk/can-utils/jsr.c b/trunk/can-utils/jsr.c
index 0303af0..baabc16 100644
--- a/trunk/can-utils/jsr.c
+++ b/trunk/can-utils/jsr.c
@@ -17,7 +17,6 @@
 
 #include <unistd.h>
 #include <getopt.h>
-#include <error.h>
 #include <poll.h>
 #include <sys/types.h>
 #include <sys/socket.h>
diff --git a/trunk/can-utils/libj1939.c b/trunk/can-utils/libj1939.c
index dc51582..8ab955c 100644
--- a/trunk/can-utils/libj1939.c
+++ b/trunk/can-utils/libj1939.c
@@ -15,7 +15,6 @@
 #include <errno.h>
 #include <inttypes.h>
 
-#include <error.h>
 #include <unistd.h>
 #include <fcntl.h>
 #include <net/if.h>
diff --git a/trunk/can-utils/testj1939.c b/trunk/can-utils/testj1939.c
index 06ca9a9..9638c08 100644
--- a/trunk/can-utils/testj1939.c
+++ b/trunk/can-utils/testj1939.c
@@ -19,7 +19,6 @@
 
 #include <unistd.h>
 #include <getopt.h>
-#include <error.h>
 #include <sys/time.h>
 #include <sys/socket.h>
 #include <net/if.h>

