--- a/pickle-temp/src/ftdi-bb.c
+++ b/pickle-temp/src/ftdi-bb.c
@@ -84,7 +84,6 @@ ftdi_bb_open(const char *device)
 			return -1;
 		}
 	}
-
 	/* all output */
 	actual_mask = 0xff;
 	if (ftdi_set_bitmode(&ftdi, actual_mask, BITMODE_SYNCBB) < 0) {
@@ -173,7 +172,7 @@ ftdi_bb_shift(struct ftdi_bb_shift *shift)
 {
 #ifdef __linux
 	uint32_t index = 0;
-	uint8_t mask;
+	uint8_t mask, pattern;
 	uint64_t value, value_mask;
 	int ret;
 	value = shift->bits;
@@ -199,22 +198,17 @@ ftdi_bb_shift(struct ftdi_bb_shift *shift)
 	value_mask = msb_first ? 1 << (shift->nbits - 1) : 1;
 
 	for (int i = 0; i< shift->nbits; i++) {
-		ftdi_buf_out[index] = pin_state;
 		if (value & value_mask)
-			ftdi_buf_out[index] |= 1 << data_pin_output | ( 1 << clock_pin);
+			pattern = pin_state | (1 << data_pin_output | 1 << clock_pin);
 		else
-			ftdi_buf_out[index] |= ( 1 << clock_pin);
+			pattern = pin_state | 1 << clock_pin;
 
-		index++;
-		ftdi_buf_out[index] = ftdi_buf_out[index-1] & ~( 1 << clock_pin);
-		index++;
-		value_mask = msb_first ? (value_mask >> 1) : (value_mask << 1);
-	}
+		ftdi_buf_out[index++] = pattern;
+		ftdi_buf_out[index++] = pattern;
+		ftdi_buf_out[index++] = pattern & ~( 1 << clock_pin);
+		ftdi_buf_out[index++] = pattern & ~( 1 << clock_pin);
 
-	/* if last bit is high - add down bit */
-	if (shift->dir == 0) {
-		ftdi_buf_out[index] = ftdi_buf_out[index-1] & ~( 1 << clock_pin) & ~(1 << data_pin_output);
-		index++;
+		value_mask = msb_first ? (value_mask >> 1) : (value_mask << 1);
 	}
 
 	if ((ret = ftdi_write_data(&ftdi, ftdi_buf_out, index )) < 0) {
@@ -225,11 +219,12 @@ ftdi_bb_shift(struct ftdi_bb_shift *shift)
 		printf("%s: ftdi_read_error [%s]\n", __func__, ftdi_get_error_string(&ftdi));
 		return -1;
 	}
+
 	value = 0;
 	value_mask = msb_first ? 1 << (shift->nbits - 1) : 1;
 	if (shift->dir) {
 		for (int i = 0; i < shift->nbits; i++ ) {
-			if (ftdi_buf_in[i*2 + 1] & (1 << data_pin_input))
+			if (ftdi_buf_in[i*4 + 2] & (1 << data_pin_input))
 				value |= value_mask;
 			value_mask = msb_first ? (value_mask >> 1) : (value_mask << 1);
 		}
--- a/pickle-temp/src/ftdi-bb.h
+++ b/pickle-temp/src/ftdi-bb.h
@@ -29,23 +29,23 @@ struct ftdi_bb_io {
 	uint8_t mode;
 	uint8_t mask;
 	uint8_t cbus_mask;
-};
+} __attribute__((packed));
 
 struct ftdi_bb_config {
-	uint8_t clock_pin;
+	uint16_t clock_pin;
+	uint16_t data_pin_input;
+	uint16_t data_pin_output;
 	uint8_t clock_falling;
-	uint8_t data_pin_input;
-	uint8_t data_pin_output;
+	uint8_t msb_first;
 	uint8_t clock_delay_low;
 	uint8_t clock_delay_high;
-	uint8_t msb_first;
-};
+} __attribute__((packed));
 
 struct ftdi_bb_shift {
 	uint8_t dir;
 	uint8_t nbits;
 	uint64_t bits;
-};
+} __attribute__((packed));
 
 int ftdi_bb_open(const char *);
 void ftdi_bb_close(void);
