diff -Nru a/s88udp-bpi.c b/s88udp-bpi.c
--- a/s88udp-bpi.c	2017-09-07 21:53:06.000000000 +0200
+++ b/s88udp-bpi.c	2017-09-07 21:53:06.000000000 +0200
@@ -29,6 +29,11 @@
 #include <linux/can.h>
 #include "s88udp-bpi.h"
 
+#ifdef USE_MRSYSTEM
+#include <mr_ipc.h>
+#include <mr_can.h>
+#endif
+
 #define BIT(x)		(1<<x)
 #define MICRODELAY	15		/* clock frequency 1/MICRODELAY[us] */
 #define MINDELAY	2		/* min delay in usec */
@@ -116,6 +121,24 @@
     printf("\n");
 }
 
+#ifdef USE_MRSYSTEM
+int sendipc(int socket, unsigned char *netframe)
+{  MrIpcCmdType Cmd;
+   MrCs2CanDataType CanMsg;
+
+   MrEthCs2Decode(&CanMsg, (char *)netframe);
+   MrIpcInit(&Cmd);
+   MrIpcEncodeFromCan(&Cmd, &CanMsg);
+   MrIpcSetSenderSocket(&Cmd, socket);
+   MrIpcSetReceiverSocket(&Cmd, MR_IPC_SOCKET_ALL);
+   if (MrIpcGetCanHash(&Cmd) == 0)
+      MrIpcSetCanHash(&Cmd, MrCs2CalcHash(MrIpcGetCanHash(&Cmd)));
+   MrIpcSend(socket, &Cmd);
+   MrIpcExit(&Cmd);
+   return(13);
+}
+#endif
+
 int create_event(struct s88_t *s88, int bus, int offset, uint32_t changed_bits, uint32_t value) {
     int s;
     uint32_t i, mask, canid, temp32;
@@ -123,7 +146,7 @@
     uint8_t netframe[13];
 
     /* allow usefull M*rklin hashes only */
-    canid = 0x00230300 | (s88->hash & 0x0000ffff);
+    canid = 0x00220300 | (s88->hash & 0x0000ffff);
 
     temp32 = htonl(canid);
     memcpy(netframe, &temp32, 4);
@@ -148,7 +171,11 @@
 		netframe[9] = 1;
 		netframe[10] = 0;
 	    }
+#ifdef USE_MRSYSTEM
+	    s = sendipc(s88->socket, netframe);
+#else
 	    s = sendto(s88->socket, netframe, 13, 0, (struct sockaddr *)&s88->baddr, sizeof(s88->baddr));
+#endif
 	    if (s != 13) {
 		fprintf(stderr, "%s: error sending UDP data: %s\n", __func__, strerror(errno));
 		return -1;
@@ -336,14 +363,20 @@
     if (!s88_data.background && s88_data.verbose)
 	printf("using broadcast address %s\n", udp_dst_address);
     /* open udp socket */
+#ifdef USE_MRSYSTEM
+    if ((s88_data.socket = MrIpcConnectIf("lo0", 10234)) < 0) {
+#else
     if ((s88_data.socket = socket(AF_INET, SOCK_DGRAM, 0)) < 0) {
-	fprintf(stderr, "UDP socket error: %s\n", strerror(errno));
+#endif
+    	fprintf(stderr, "UDP socket error: %s\n", strerror(errno));
 	exit(EXIT_FAILURE);
     }
+#ifndef USE_MRSYSTEM
     if (setsockopt(s88_data.socket, SOL_SOCKET, SO_BROADCAST, &on, sizeof(on)) < 0) {
 	fprintf(stderr, "UDP set broadcast option error: %s\n", strerror(errno));
 	exit(EXIT_FAILURE);
     }
+#endif
 
     if (destination_second_port) {
 	/* open second udp socket if requested */
