diff --git a/can2udp/debian/changelog b/can2udp/debian/changelog
index 6d08518..6b51a74 100644
--- a/can2udp/debian/changelog
+++ b/can2udp/debian/changelog
@@ -1,5 +1,6 @@
-can2udp (0.0.git20171130) UNRELEASED; urgency=medium
+can2udp (0.0.git20171229) UNRELEASED; urgency=medium
 
   * Debianised
+  * include s88can
 
- -- Joerg Dorchain <joerg@Redstar.dorchain.net>  Thu, 30 Nov 2017 10:06:12 +0100
+ -- Joerg Dorchain <joerg@Redstar.dorchain.net>  Sun, 18 Feb 2018 09:07:52 +0100
diff --git a/can2udp/debian/rules b/can2udp/debian/rules
index 08d05a5..2cf1f55 100755
--- a/can2udp/debian/rules
+++ b/can2udp/debian/rules
@@ -9,7 +9,7 @@ override_dh_auto_clean:
     dh_clean
 
 override_dh_auto_build:
-    cd src; make can2lan can2serial lan-schnitte can-monitor cs2ping gb2-update pio-i2c-udp s88udp-rpi s88udp-bpi
+    cd src; CPPFLAGS=-DSUNXI_A20 make can2lan can2serial lan-schnitte can-monitor cs2ping gb2-update pio-i2c-udp s88udp-rpi s88udp-bpi s88can
 
 override_dh_auto_install:
     cp src/can2lan $$(pwd)/debian/can2lan/usr/sbin
@@ -21,6 +21,7 @@ override_dh_auto_install:
     cp src/pio-i2c-udp $$(pwd)/debian/can2udp-utils/usr/bin
     cp src/s88udp-bpi $$(pwd)/debian/can2udp-utils/usr/bin
     cp src/s88udp-rpi $$(pwd)/debian/can2udp-utils/usr/bin
+    cp src/s88can $$(pwd)/debian/can2udp-utils/usr/bin
 
 override_dh_shlibdeps:
     dh_shlibdeps --dpkg-shlibdeps-params=--ignore-missing-info
--- a/can2udp/src/s88can.c.orig    2018-02-18 09:07:15.538692795 +0100
+++ b/can2udp/src/s88can.c    2018-02-18 09:07:19.642610596 +0100
@@ -352,9 +352,11 @@
 memset(s88[config_num - 3].bus_ct1, 0xff, sizeof(s88[config_num - 3].bus_ct1));
 s88[config_num - 3].poll = bpi_poll;
 
-for( i = 0, tp = strtok(lenp, ","); (tp != NULL) && (i < config_num - 2); i++) {
-  s88[i].blen = atoi(tp);
-  tp = strtok(NULL, ",");
+if (lenp != NULL) {
+  for( i = 0, tp = strtok(lenp, ","); (tp != NULL) && (i < config_num - 2); i++) {
+    s88[i].blen = atoi(tp);
+    tp = strtok(NULL, ",");
+  }
 }
 
 if (!foreground) {
