--- a/makefile
+++ b/makefile
@@ -72,6 +72,7 @@
 
 version:
 	echo $(QUOT)const int bzr = $(QUOT) > common$(FS)version.h
-	bzr revno --tree >> common$(FS)version.h
+#	bzr revno --tree >> common$(FS)version.h
+	echo $(QUOT)6318$(QUOT) >> common$(FS)version.h
 	echo $(QUOT);$(QUOT) >> common$(FS)version.h
 
--- a/rocint/makefile
+++ b/rocint/makefile
@@ -76,8 +76,8 @@ TMPOUTDIR=$(GENDIR)$(FS)rocint$(FS)bin
 DEBUG=-g
 BINDIR=..$(FS)$(COREDIR)bin
 
-CPP=gcc
-LNK=gcc
+CPP=$(TOOLPREFIX)gcc
+LNK=$(TOOLPREFIX)gcc
 INCL_PATH=$(MOUNTPOINT)
 CC_FLAGS=-c $(CC_EXTRA_FLAGS) $(DEBUG) -ansi -I$(INCL_PATH) -I$(GENDIR)
 
--- a/roclcdr/makefile
+++ b/roclcdr/makefile
@@ -86,7 +86,7 @@ else
 	DIRPREFIX=unx
 	CC_EXTRA_FLAGS=-fPIC
 	LNK_FLAGS=-shared
-	LNK=gcc
+	LNK=$(TOOLPREFIX)gcc
 	SHAREDSUFFIX=.so
 endif
 
--- a/rocdigs/makefile
+++ b/rocdigs/makefile
@@ -86,7 +86,7 @@ else
 	DIRPREFIX=unx
 	CC_EXTRA_FLAGS=-fPIC
 	LNK_FLAGS=-shared
-	LNK=gcc
+	LNK=$(TOOLPREFIX)gcc
 	SHAREDSUFFIX=.so
 endif

--- a/rocnetnode/makefile
+++ b/rocnetnode/makefile
@@ -131,6 +131,7 @@
 
 version:
 	echo $(QUOT)const int bzr = $(QUOT) > ..$(FS)common$(FS)version.h
-	bzr revno --tree >> ..$(FS)common$(FS)version.h
+#	bzr revno --tree >> ..$(FS)common$(FS)version.h
+	echo $(QUOT)6318$(QUOT) >> ..$(FS)common$(FS)version.h
 	echo $(QUOT);$(QUOT) >> ..$(FS)common$(FS)version.h
 
