--- a/rocs/impl/unx/userial.c
+++ b/rocs/impl/unx/userial.c
@@ -197,7 +197,7 @@ Boolean rocs_serial_open( iOSerial inst
       o->portbase = 0x2E8;
   }

-#if defined __APPLE__ || defined __OpenBSD__
+#if defined __APPLE__ || defined __OpenBSD__ || defined __NO_DIRECT_IO__
     o->directIO=False;
 #else
   if( o->portbase > 0 && ioperm(o->portbase, 8, 1) == 0 ) {
@@ -351,7 +351,7 @@ static void __printmsr(int msr) {
 static int __last_val = 0;
 static Boolean __printport(iOSerial inst, int* richg) {
   iOSerialData o = Data(inst);
-#if defined __APPLE__ || defined __OpenBSD__
+#if defined __APPLE__ || defined __OpenBSD__ || defined __NO_DIRECT_IO__
     return True;
 #else
   if( ioperm(o->portbase, 7, 1) == 0 ) {
@@ -503,9 +503,11 @@ Boolean rocs_serial_isRI( iOSerial inst
   iOSerialData o = Data(inst);
   int msr = 0;
   int result, arg;
+#if !defined __NO_DIRECT_IO__
   if (o->directIO) {
     msr = inb(o->portbase+6);
   }
+#endif
   result = ioctl(o->sh, TIOCMGET, &arg);
   if ((result>=0)&&((!(arg&TIOCM_RI))||(msr&0x04)))
     return True;
@@ -606,7 +608,9 @@ Boolean rocs_serial_isUartEmpty( iOSeria
     } else
       /* TIOCSERGETLSR worked, return result*/
       return result ? True:False;
-  } else if( ioperm(o->portbase, 7, 1) == 0 ) {
+  }
+#if !defined __NO_DIRECT_IO__
+   else if( ioperm(o->portbase, 7, 1) == 0 ) {
     int lsr = inb(o->portbase+5);
     /*UART is empty*/
     if(lsr&0x40)
@@ -614,6 +618,7 @@ Boolean rocs_serial_isUartEmpty( iOSeria
     else
       return False;
   } else
+#endif
     return True;
 #endif
 #endif
