--- a/rocs/impl/unx/uthread.c
+++ b/rocs/impl/unx/uthread.c
@@ -128,7 +128,7 @@
     else {
       TraceOp.terrno( name, TRCLEVEL_WARNING, __LINE__, 9999, rc, "Could not set thread priority to %d (only superuser can do this)", param.sched_priority );
     }
-    rc = pthread_getschedparam( o->handle, &policy, &param);
+    rc = pthread_getschedparam( (pthread_t)o->handle, &policy, &param);
     if( policy == SCHED_FIFO ) szPolicy = "SCHED_FIFO";
     else if( policy == SCHED_RR ) szPolicy = "SCHED_RR";
     else szPolicy = "SCHED_OTHER";
