diff --git a/rocnetnode/makefile b/rocnetnode/makefile
index 3e533d8..7150b3d 100644
--- a/rocnetnode/makefile
+++ b/rocnetnode/makefile
@@ -128,7 +128,7 @@ clean:
	-$(REMOVE) $(OUTDIR)$(FS)*.o
	-$(REMOVE) $(GENDIR)$(FS)rocnetnode$(FS)bin$(FS)*
	-$(REMOVE) $(GENDIR)$(FS)rocnetnode$(FS)public$(FS)*
-	cd $(GENDIR)$(FS)rocnetnode$(CS) ..$(FS)$(BINDIR)$(FS)ogen$(GENSUFFIX) ..$(FS)..$(FS)rocnetnode$(FS)rocnetnode.xml ..$(FS)
+	cd $(GENDIR)$(FS)rocnetnode$(CS) ..$(FS)..$(FS)rocs$(FS)bin$(FS)ogen$(GENSUFFIX)  ..$(FS)..$(FS)rocnetnode$(FS)rocnetnode.xml ..$(FS)


 install:

