diff --git a/feeds.conf.default b/feeds.conf.default
index da6c665624..4999f040c9 100644
--- a/feeds.conf.default
+++ b/feeds.conf.default
@@ -1,9 +1,10 @@
 src-git packages https://git.openwrt.org/feed/packages.git
 src-git luci https://git.openwrt.org/project/luci.git
 src-git routing https://git.openwrt.org/feed/routing.git
-src-git telephony https://git.openwrt.org/feed/telephony.git
+#src-git telephony https://git.openwrt.org/feed/telephony.git
 #src-git video https://github.com/openwrt/video.git
 #src-git targets https://github.com/openwrt/targets.git
 #src-git management https://github.com/openwrt-management/packages.git
 #src-git oldpackages http://git.openwrt.org/packages.git
 #src-link custom /usr/src/openwrt/custom-feed
+src-git openwrtmisc https://github.com/GBert/openwrt-misc
diff --git a/package/utils/adb/patches/010-sysmacro-fix.patch b/package/utils/adb/patches/010-sysmacro-fix.patch
new file mode 100644
index 0000000000..ad2630ea54
--- /dev/null
+++ b/package/utils/adb/patches/010-sysmacro-fix.patch
@@ -0,0 +1,24 @@
+Index: adb-6fe92d1a3fb17545d82d020a3c995f32e6b71f9d/adb/usb_linux.c
+===================================================================
+--- adb-6fe92d1a3fb17545d82d020a3c995f32e6b71f9d.orig/adb/usb_linux.c
++++ adb-6fe92d1a3fb17545d82d020a3c995f32e6b71f9d/adb/usb_linux.c
+@@ -21,6 +21,7 @@
+ 
+ #include <sys/ioctl.h>
+ #include <sys/types.h>
++#include <sys/sysmacros.h>
+ #include <sys/time.h>
+ #include <dirent.h>
+ #include <fcntl.h>
+Index: adb-6fe92d1a3fb17545d82d020a3c995f32e6b71f9d/fastboot/usb_linux.c
+===================================================================
+--- adb-6fe92d1a3fb17545d82d020a3c995f32e6b71f9d.orig/fastboot/usb_linux.c
++++ adb-6fe92d1a3fb17545d82d020a3c995f32e6b71f9d/fastboot/usb_linux.c
+@@ -34,6 +34,7 @@
+ #include <sys/ioctl.h>
+ #include <sys/stat.h>
+ #include <sys/types.h>
++#include <sys/sysmacros.h>
+ #include <dirent.h>
+ #include <fcntl.h>
+ #include <errno.h>
diff --git a/target/linux/ramips/base-files/etc/board.d/02_network b/target/linux/ramips/base-files/etc/board.d/02_network
index 2f9a02256e..975c3a0dfc 100755
--- a/target/linux/ramips/base-files/etc/board.d/02_network
+++ b/target/linux/ramips/base-files/etc/board.d/02_network
@@ -165,6 +165,8 @@ ramips_setup_interfaces()
 	thunder,timecloud|\
 	tplink,tl-wa801nd-v5|\
 	unbranded,a5-v11|\
+	unbranded,a5-v11-8m|\
+	unbranded,a5-v11-16m|\
 	wansview,ncs601w|\
 	widora,neo-16m|\
 	widora,neo-32m|\
diff --git a/target/linux/ramips/dts/rt5350_unbranded_a5-v11-16m.dts b/target/linux/ramips/dts/rt5350_unbranded_a5-v11-16m.dts
new file mode 100644
index 0000000000..1044df50cb
--- /dev/null
+++ b/target/linux/ramips/dts/rt5350_unbranded_a5-v11-16m.dts
@@ -0,0 +1,134 @@
+/dts-v1/;
+
+#include "rt5350.dtsi"
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/input.h>
+
+/ {
+	compatible = "unbranded,a5-v11", "ralink,rt5350-soc";
+	model = "A5-V11";
+
+	aliases {
+		led-boot = &led_power;
+		led-failsafe = &led_power;
+		led-running = &led_power;
+		led-upgrade = &led_power;
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		system {
+			label = "a5-v11:blue:system";
+			gpios = <&gpio0 20 GPIO_ACTIVE_LOW>;
+		};
+
+		led_power: power {
+			label = "a5-v11:red:power";
+			gpios = <&gpio0 17 GPIO_ACTIVE_LOW>;
+		};
+	};
+
+	keys {
+		compatible = "gpio-keys-polled";
+		poll-interval = <20>;
+
+		reset {
+			label = "reset";
+			gpios = <&gpio0 0 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_RESTART>;
+		};
+	};
+
+	gpio_export {
+		compatible = "gpio-export";
+		#size-cells = <0>;
+
+		usb {
+			gpio-export,name = "usb";
+			gpio-export,output = <1>;
+			gpios = <&gpio0 7 GPIO_ACTIVE_HIGH>;
+		};
+
+		root_hub {
+			gpio-export,name = "root_hub";
+			gpio-export,output = <1>;
+			gpios = <&gpio0 12 GPIO_ACTIVE_HIGH>;
+		};
+	};
+};
+
+&spi0 {
+	status = "okay";
+
+	m25p80@0 {
+		compatible = "jedec,spi-nor";
+		reg = <0>;
+		spi-max-frequency = <10000000>;
+
+		partitions {
+			compatible = "fixed-partitions";
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			partition@0 {
+				label = "u-boot";
+				reg = <0x0 0x30000>;
+				read-only;
+			};
+
+			partition@30000 {
+				label = "u-boot-env";
+				reg = <0x30000 0x10000>;
+				read-only;
+			};
+
+			factory: partition@40000 {
+				label = "factory";
+				reg = <0x40000 0x10000>;
+				read-only;
+			};
+
+			partition@50000 {
+				compatible = "denx,uimage";
+				label = "firmware";
+				reg = <0x50000 0xfb0000>;
+			};
+		};
+	};
+};
+
+&gpio1 {
+	status = "okay";
+};
+
+&pinctrl {
+	state_default: pinctrl0 {
+		gpio {
+			ralink,group = "i2c", "jtag", "uartf", "led";
+			ralink,function = "gpio";
+		};
+	};
+};
+
+&ethernet {
+	mtd-mac-address = <&factory 0x4>;
+};
+
+&esw {
+	mediatek,portmap = <0x1>;
+	mediatek,portdisable = <0x3e>;
+};
+
+&wmac {
+	ralink,mtd-eeprom = <&factory 0>;
+};
+
+&ehci {
+	status = "okay";
+};
+
+&ohci {
+	status = "okay";
+};
diff --git a/target/linux/ramips/dts/rt5350_unbranded_a5-v11-8m.dts b/target/linux/ramips/dts/rt5350_unbranded_a5-v11-8m.dts
new file mode 100644
index 0000000000..8b4ec29ef8
--- /dev/null
+++ b/target/linux/ramips/dts/rt5350_unbranded_a5-v11-8m.dts
@@ -0,0 +1,134 @@
+/dts-v1/;
+
+#include "rt5350.dtsi"
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/input.h>
+
+/ {
+	compatible = "unbranded,a5-v11", "ralink,rt5350-soc";
+	model = "A5-V11";
+
+	aliases {
+		led-boot = &led_power;
+		led-failsafe = &led_power;
+		led-running = &led_power;
+		led-upgrade = &led_power;
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		system {
+			label = "a5-v11:blue:system";
+			gpios = <&gpio0 20 GPIO_ACTIVE_LOW>;
+		};
+
+		led_power: power {
+			label = "a5-v11:red:power";
+			gpios = <&gpio0 17 GPIO_ACTIVE_LOW>;
+		};
+	};
+
+	keys {
+		compatible = "gpio-keys-polled";
+		poll-interval = <20>;
+
+		reset {
+			label = "reset";
+			gpios = <&gpio0 0 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_RESTART>;
+		};
+	};
+
+	gpio_export {
+		compatible = "gpio-export";
+		#size-cells = <0>;
+
+		usb {
+			gpio-export,name = "usb";
+			gpio-export,output = <1>;
+			gpios = <&gpio0 7 GPIO_ACTIVE_HIGH>;
+		};
+
+		root_hub {
+			gpio-export,name = "root_hub";
+			gpio-export,output = <1>;
+			gpios = <&gpio0 12 GPIO_ACTIVE_HIGH>;
+		};
+	};
+};
+
+&spi0 {
+	status = "okay";
+
+	m25p80@0 {
+		compatible = "jedec,spi-nor";
+		reg = <0>;
+		spi-max-frequency = <10000000>;
+
+		partitions {
+			compatible = "fixed-partitions";
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			partition@0 {
+				label = "u-boot";
+				reg = <0x0 0x30000>;
+				read-only;
+			};
+
+			partition@30000 {
+				label = "u-boot-env";
+				reg = <0x30000 0x10000>;
+				read-only;
+			};
+
+			factory: partition@40000 {
+				label = "factory";
+				reg = <0x40000 0x10000>;
+				read-only;
+			};
+
+			partition@50000 {
+				compatible = "denx,uimage";
+				label = "firmware";
+				reg = <0x50000 0x7b0000>;
+			};
+		};
+	};
+};
+
+&gpio1 {
+	status = "okay";
+};
+
+&pinctrl {
+	state_default: pinctrl0 {
+		gpio {
+			ralink,group = "i2c", "jtag", "uartf", "led";
+			ralink,function = "gpio";
+		};
+	};
+};
+
+&ethernet {
+	mtd-mac-address = <&factory 0x4>;
+};
+
+&esw {
+	mediatek,portmap = <0x1>;
+	mediatek,portdisable = <0x3e>;
+};
+
+&wmac {
+	ralink,mtd-eeprom = <&factory 0>;
+};
+
+&ehci {
+	status = "okay";
+};
+
+&ohci {
+	status = "okay";
+};
diff --git a/target/linux/ramips/image/rt305x.mk b/target/linux/ramips/image/rt305x.mk
index 0a03449449..987b807a5c 100644
--- a/target/linux/ramips/image/rt305x.mk
+++ b/target/linux/ramips/image/rt305x.mk
@@ -1045,6 +1045,32 @@ define Device/unbranded_a5-v11
 endef
 TARGET_DEVICES += unbranded_a5-v11
 
+define Device/unbranded_a5-v11-8m
+  MTK_SOC := rt5350
+  IMAGE_SIZE := 7872k
+  IMAGES += factory.bin
+  IMAGE/factory.bin := \
+       $$(sysupgrade_bin) | check-size $$$$(IMAGE_SIZE) | poray-header -B A5-V11 -F 8M
+  DEVICE_VENDOR :=
+  DEVICE_TITLE := A5-V11 (8M)
+  DEVICE_PACKAGES := kmod-usb-core kmod-usb-ohci kmod-usb2
+  SUPPORTED_DEVICES += a5-v11
+endef
+TARGET_DEVICES += unbranded_a5-v11-8m
+
+define Device/unbranded_a5-v11-16m
+  MTK_SOC := rt5350
+  IMAGE_SIZE := 16064k
+  IMAGES += factory.bin
+  IMAGE/factory.bin := \
+       $$(sysupgrade_bin) | check-size $$$$(IMAGE_SIZE) | poray-header -B A5-V11 -F 16M
+  DEVICE_VENDOR :=
+  DEVICE_TITLE := A5-V11 (16M)
+  DEVICE_PACKAGES := kmod-usb-core kmod-usb-ohci kmod-usb2
+  SUPPORTED_DEVICES += a5-v11
+endef
+TARGET_DEVICES += unbranded_a5-v11-16m
+
 define Device/unbranded_wr512-3gn-4m
   MTK_SOC := rt3052
   IMAGE_SIZE := 3776k
diff --git a/tools/firmware-utils/src/mkporayfw.c b/tools/firmware-utils/src/mkporayfw.c
index 6ec4f320d9..4497e0bca0 100644
--- a/tools/firmware-utils/src/mkporayfw.c
+++ b/tools/firmware-utils/src/mkporayfw.c
@@ -135,6 +135,9 @@ static struct flash_layout layouts[] = {
 	}, {
 		.id		= "8M",
 		.fw_max_len	= 0x7c0000,
+	}, {
+		.id		= "16M",
+		.fw_max_len	= 0xfc0000,
 	}, {
 		/* terminating entry */
 	}
